pure-wasm/src/lib.rs

```diff
@@ -191,6 +191,22 @@ pub extern "C" fn take_state(engine_ptr: u32, json_ptr: u32, json_len: u32) {
     }
 }
 
+#[unsafe(no_mangle)]
+pub extern "C" fn get_state(engine_ptr: u32) -> u32 {
+    let lock = unsafe { get_engine(engine_ptr as *const u32).unwrap() };
+    let engine = recover_lock(&lock);
+
+    let state = engine.get_state();
+    if let Ok(json_str) = serde_json::to_string(&state) {
+        let json_bytes = json_str.as_bytes();
+        let json_ptr = json_bytes.as_ptr() as u32;
+        std::mem::forget(json_str); // Prevent automatic deallocation
```

Comment:
This function intentionally leaks memory (the JSON string) to make it available to the calling code.